### PR TITLE
feat: handle labeled dataset

### PR DESCRIPTION
**Pull request by JulesBelveze**
This PR aims at enhancing the existing project with the two following features:
- [x] handle extra columns in the dataframe (e.g. metadata)
- [x] use and display labels if existing in the dataframe

The idea here is to be able to provide more data points information to the user, enabling he/she to discover underlying patterns in the data.

### CODE REVIEW

**Comment by koaning** [comment]
I like the idea, and I can see a general benefit but there's something else to consider. What if the "labels" column is numeric? Then we might want to allow for something like this: 

image


What if the column contains "NaN"? Maybe then we'd want something like this: 

image

Note that these are images from [cluestar](https://github.com/koaning/cluestar); it's a related package that's only supposed to make pretty renderings inside Jupyter. It's kind of meant as a tool that you can use as "the work before sending data into bulk". Bulk is similar, but it is supposed to allow you to label in a separate environment.

Maybe it'd be better to have a "color" column that has the behavior that I just described? Calling it "label" is a bit weird when you consider that bulk is supposed to actually help you label. If the label already exists, there's no need for bulk.

**Review by koaning — CHANGES_REQUESTED** [review]
Let's discuss the API/design/behavior a bit more for this feature.

**Comment by koaning** [comment]
Given that folks are making PRs now. I'm also wondering if it makes sense to think about some unit tests. These are usually very hard for visualisation tools, but this might be a nice use-case for [playwright](https://playwright.dev/). No worries about setting this up though, that'll be my responsibility as maintainer, but if you have any ideas I'm all ears.

**Comment by JulesBelveze** [comment]
I actually used `bulk` for a slightly different use case: to be able to spot wrongly labeled data (or duplicates) and then update the labels (or delete samples) in the annotation tool I used.

What you are saying about the naming of the column and the data type makes total sense. I'll make changes.

Also, I think it could be interesting for the user to select any column of the dataset and colorize the data points based on it. What do you think?

**Comment by koaning** [comment]
> Also, I think it could be interesting for the user to select any column of the dataset and colorize the data points based on it. What do you think?

I would save that for another PR. I want to prevent scope creep, and adding more interactivity might also make it harder to maintain. Does Bokeh even allow for column selection events to trigger an update?

**Comment by koaning** [comment]
Also, if you're looking for wrongly labelled data ... have you seen [doubtlab](https://github.com/koaning/doubtlab), my [calmcode tutorial](https://calmcode.io/bad-labels/introduction.html) or my [Prodigy tutorial](https://www.youtube.com/watch?v=khZ5-AN-n2Y)? 😅

**Comment by JulesBelveze** [comment]
> I would save that for another PR. I want to prevent scope creep, and adding more interactivity might also make it harder to maintain.

Yeah agree, let us stick to the original idea. I'll make the needed changes and ping you once done 😺 

> Does Bokeh even allow for column selection events to trigger an update?

Quite unsure, I've actually never seen it in the past 😅

> Also, if you're looking for wrongly labelled data ... have you seen [doubtlab](https://github.com/koaning/doubtlab), my [calmcode tutorial](https://calmcode.io/bad-labels/introduction.html) or my [Prodigy tutorial](https://www.youtube.com/watch?v=khZ5-AN-n2Y)? 😅

Hehe, yes I have! I do use either `doubtlab` or [programmatic](https://docs.programmatic.humanloop.com/overview/readme) for quality insurance of annotations.

**Comment by koaning** [comment]
I didn't know about programmatic! Have just made a mental note to check it out though.

**Comment by koaning** [comment]
If you have any feedback for doubtlab, feel free to leave a comment on the issue list btw.

**Comment by JulesBelveze** [comment]
@koaning let me know what you think! I've tried it for different "color" value types as suggested. If that's fine by you, I'll resolve the merge conflicts

File: setup.py
Comment (koaning) [inline setup.py]: Why is matplotlib a required dependency now?

File: bulk/__main__.py
Comment (koaning) [inline bulk/__main__.py]: This version implementation is currently wrong, it needs to pick the actual version from the package. I'll fix that later, but it's preferable to leave this untouched. That might just prevent a few merge conflicts.

File: setup.py
Comment (JulesBelveze) [reply setup.py]: My mistake

File: bulk/__main__.py
Comment (JulesBelveze) [reply bulk/__main__.py]: Okay 😃

File: bulk/text.py
Comment (koaning) [inline bulk/text.py]: Before my assumption was that folks are only interested in storing the text. You seem to want to include more. That can be fine, but why do you assume that folks wouldn't be interested in the `x`, `y` and `color` columns?

**Review by koaning — CHANGES_REQUESTED** [review]
I still need to run this locally, but left a few comments. 

Two extra things:

1. I'm cool with adding an image that demonstrates the color, but I'd prefer the first image to be simple. Could you also add an image that has less whitespace in it? 
2. Is there a reason why the `cluestarred.csv` file is now smaller than before?

File: bulk/text.py
Comment (JulesBelveze) [reply bulk/text.py]: Well, the `x` and `y` columns are pretty hard to interpret if not plotted. One might be interested in the `color` though. But I agree it is debatable whether to save only the `text` column or also the metadata associated with a data point.
Should we agree on saving all the columns but `x` and `y`?

**Comment by JulesBelveze** [comment]
> I'm cool with adding an image that demonstrates the color, but I'd prefer the first image to be simple. Could you also add an image that has less whitespace in it?

@koaning I have cropped the whitespace of the first image.

> 2. Is there a reason why the cluestarred.csv file is now smaller than before?

Also, I rolled back `cluestarred.csv` to its original version.

Let me know if you think we should add an image with the color.

File: bulk/text.py
Comment (koaning) [reply bulk/text.py]: Technically, the tool is supposed to merely make a subset of the data. So that suggests to me that the most "basic" behaviour is just to do a subset and nothing else. That would mean we usually just include all variables. It's very little effort to filter them away later and I might prefer moving that responsibility to the end-user if that means we don't have to make an assumption. 

I do think that it'd be nice to have a `--just-text` flag though. But that's perhaps out of scope for this PR. 

For this PR, would you object to just returning all columns?

Comment (JulesBelveze) [reply bulk/text.py]: Agree with you it would be nice to have a flag!
Sure, I'll make the changes to return all columns 😃

Comment (JulesBelveze) [reply bulk/text.py]: Done in be48a51

**Comment by koaning** [comment]
Thanks for the PRs! I'll give it another, more in-depth, look later today.

**Comment by koaning** [comment]
I was able to break it by first generating this: 

```python
df = pd.read_csv("cluestarred.csv")
df.assign(color=lambda d: d['text'].str.contains("deliver").astype(str)).to_csv("colored.csv")
```

When I pass this `colored.csv` file I get a status 500 from the `bulk` command. It comes with this traceback. 

```
ERROR:tornado.application:Uncaught exception GET / (::1)
HTTPServerRequest(protocol='http', host='localhost:5006', method='GET', uri='/', version='HTTP/1.1', remote_ip='::1')
Traceback (most recent call last):
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/tornado/web.py", line 1704, in _execute
    result = await result
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/server/views/doc_handler.py", line 54, in get
    session = await self.get_session()
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/server/views/session_handler.py", line 144, in get_session
    session = await self.application_context.create_session_if_needed(session_id, self.request, token)
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/server/contexts.py", line 243, in create_session_if_needed
    self._application.initialize_document(doc)
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/application/application.py", line 194, in initialize_document
    h.modify_document(doc)
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/application/handlers/function.py", line 143, in modify_document
    self._func(doc)
  File "/Users/vincentwarmerdam/Development/bulk/bulk/text.py", line 43, in bkapp
    mapper = get_color_mapping(df)
  File "/Users/vincentwarmerdam/Development/bulk/bulk/utils.py", line 15, in get_color_mapping
    mapper = factor_cmap(
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/transform.py", line 144, in factor_cmap
    return field(field_name, CategoricalColorMapper(palette=palette,
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/models/mappers.py", line 98, in __init__
    super().__init__(**kwargs)
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/model/model.py", line 128, in __init__
    super().__init__(**kwargs)
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/core/has_props.py", line 206, in __init__
    setattr(self, name, value)
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/core/has_props.py", line 230, in __setattr__
    return super().__setattr__(name, value)
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/core/property/descriptors.py", line 283, in __set__
    value = self.property.prepare_value(obj, self.name, value)
  File "/Users/vincentwarmerdam/Development/bulk/venv/lib/python3.8/site-packages/bokeh/core/property/bases.py", line 365, in prepare_value
    raise ValueError(f"failed to validate {obj_repr}.{name}: {error}")
ValueError: failed to validate CategoricalColorMapper(id='1032', ...).factors: expected an element of either Seq(String), Seq(Tuple(String, String)) or Seq(Tuple(String, String, String)), got [False, True, nan]
ERROR:tornado.access:500 GET / (::1) 53.55ms
```

I would expect that the `color` column would be able to also handle boolean values. Does this also happen on your end?

**Comment by koaning** [comment]
There's another thing when I try putting in pure strings. 

```python
import random

df.assign(color=lambda d: ['a' if random.random() < 0.5 else 'b' for _ in range(d.shape[0])]).to_csv("colored.csv")
```

image

Notice how the unnamed column also is shown in the interface? I think it's best to remove it since it moves the text too far to the right. 

I'm also wondering if there's another color palette we could pick. The yellow is somewhat hard to see.

**Comment by koaning** [comment]
I can confirm that it's able to support floating values. Nice!

**Comment by JulesBelveze** [comment]
The boolean problem should be fixed by a628c09. I hope the color palette is better, I could not find any great fit..
Let me know if there's anything else to change 😺

File: bulk/text.py
Comment (koaning) [inline bulk/text.py]: Ah! So that's why the points feel much smaller than before. 

While I wouldn't mind giving the user some way to change the styles of the chart (maybe via a config file of sorts?) I do feel that this should be tackled in another PR. Can we leave the base settings as-is for now?

File: bulk/text.py
Comment (koaning) [inline bulk/text.py:16]: Because we're adding more columns now, the text truncates in the UI.

image

Because we're most interested in the text here, I'm not in favour of adding any columns that might make it harder to inspect it. Can you make sure that we only show the text column here?

**Review by koaning — CHANGES_REQUESTED** [review]
It seems like you've adapted the `clustarred.csv` file again. Could you revert that change? 

It also seems like you're still replacing the original readme image. I'll gladly add another image to the readme later, but the original image should remain "as simple as possible".

File: bulk/text.py
Comment (koaning) [reply bulk/text.py]: Another thing about this: I'll likely add a zoom widget to the chart after this PR is merged. That way you can zoom in properly if you need more space to draw the shape.

Comment (JulesBelveze) [reply bulk/text.py]: I think that the zooming feature is really relevant, I found myself struggling a bit by drawing the shape sometimes..

**Comment by JulesBelveze** [comment]
@koaning Everything should **finally** be in order now 😅 sorry for that

File: bulk/text.py
Comment (JulesBelveze) [reply bulk/text.py]: If you wanna open an issue, I can have a look at it 😃

Comment (koaning) [reply bulk/text.py]: It's just an [extra widget](https://docs.bokeh.org/en/latest/docs/user_guide/tools.html#scroll-pinch-tools) that needs to be configured for Bokeh. 

But feel free to make a PR for it 👍

Comment (JulesBelveze) [reply bulk/text.py]: Will do !

**Comment by koaning** [comment]
Just to check, can you confirm that when you run it locally that the file actually gets saved appropriately? I'm seeing empty rows on my machine when I pull this branch.

**Comment by koaning** [comment]
And the PR also still suggests that we're changing the .csv file? 

image

**Review by koaning — CHANGES_REQUESTED** [review]
Just to check, do you get proper files as output on your machine?

**Comment by JulesBelveze** [comment]
@koaning Hmmm seems like a weird typo that wasn't caught by my linter...
Does it work on your hand now?

**Comment by koaning** [comment]
I'll check it now, in the meantime though: the .csv file still needs to be changed back to the original.

**Comment by koaning** [comment]
I can confirm that it works now locally, but I do need to figure out a way to unit-test this. 

Let me know when've undone the changes in the .csv file :)

**Comment by JulesBelveze** [comment]
@koaning Sorry my linter was automatically reformating the file.. Should be good now :)

Yes, would be quite convenient to have unit-tests!

**Comment by koaning** [comment]
Once your changes are in I can add these new features:

image

**Comment by JulesBelveze** [comment]
@koaning Looks good!
I think it should be ready to be merged now

**Review by koaning — APPROVED** [review]

**Pull request merged**
